Add helper functions  to draw the instrument beam and a physical scale for WCSAxes instances

smaret opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

This pull request adds two features of the [APLPy package](https://aplpy.readthedocs.io/en/stable/) which are missing in Astropy, and that allow to:

1. Draw an ellipse representing the instrument beam on an image
2. Draw a segment that represents a physical scale (for example 100 au) on an image 

Both features as implemented as helper functions in `astropy.visualization.wcs`, `add_beam()` and `add_scalebar()`, respectively. The code of these functions is borrowed from APLPy (see [here](https://aplpy.readthedocs.io/en/stable/_modules/aplpy/overlays.html#Beam.show) and [here](https://aplpy.readthedocs.io/en/stable/api/aplpy.Scalebar.html#aplpy.Scalebar.show)).

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

smaret commented:
For reference, here are the three figures that are created by the tests:
![test_beam_shape_from_args](https://user-images.githubusercontent.com/95471/134779630-cc0d5dae-d50f-428c-bf93-013369ac8b93.png)
![test_beam_shape_from_header](https://user-images.githubusercontent.com/95471/134779537-bcb9c305-7ade-47df-a5eb-64788182e6c3.png)
![test_scalebar](https://user-images.githubusercontent.com/95471/134779542-63c363c1-6da6-485e-a1cd-6c27ec2526fe.png)

astrofrog commented:
Thanks for the PR! I will try and review this soon!

astrofrog commented:
I am re-milestoning to v5.1 as we've now branched v5.0.x - I'm generally on board with this but I think it will require a little more work. I will make some concrete suggestions shortly.

smaret commented:
@pllim I don't understand why the change log test is failing. I did add a change log entry (see 2f89a62).

pllim commented:
Sorry for the confusion, I deployed "what's new entry" check into the same job. The error: `"whatsnew-needed" present but no What's New entry; please add one.`

keflavich commented:
For both of the issues @ayshih noted, these are likely to be atypical corner cases - it's been pretty rare that we've seen rectangular pixels in FITS images in the wild.  It would be a good idea to warn the user if these cases are noticed, but correctly handling them shouldn't block including this feature.

ayshih commented:
Although rectangular pixels are not common, I think they are more common in simulated data and more common in heliophysics than astrophysics.  For example, I was Co-I an instrument concept a couple of years ago that would have imaged via rectangular collimators with an aspect ratio of ~6.

If nothing else, I suggest that such cases be noted in the docstrings.

saimn commented:
Moving the milestone to 5.2 since the PR doesn't seems ready.

smaret commented:
> I am re-milestoning to v5.1 as we've now branched v5.0.x - I'm generally on board with this but I think it will require a little more work. I will make some concrete suggestions shortly.

@astrofrog Any update on this?

astrofrog commented:
I am mostly off work this month but will get back to you in July!

astrofrog commented:
I plan to get to this in the new few days now that other WCSAxes pull requests have been merged.

astrofrog commented on astropy/visualization/wcsaxes/core.py:
```diff
@@ -96,19 +108,28 @@ class WCSAxes(Axes):
+        frame_class=None,
+        **kwargs,
+    ):
+        """ """
```
We are still discussing whether or not to adopt black formatting for astropy (see https://github.com/astropy/astropy-APEs/pull/76) - since this hasn't been approved yet (and may not be), could you undo any formatting changes for now? (alternatively you could always wait a few weeks to see if the APE is approved, if undoing the formatting changes would be too much work).

smaret replied:
OK, done.

astrofrog commented on astropy/visualization/wcsaxes/core.py:
```diff
@@ -307,25 +332,179 @@ def plot_coord(self, *args, **kwargs):
+        borderpad=0.4,
+        pad=0.5,
+        **kwargs,
+    ):
```
I would prefer to keep these as helper functions that don't live as methods on the class, because they only apply to specific kinds of data (for instance only celestial WCS) - I think we could develop a whole set of helper functions for different kinds of WCS and would prefer not to add too much to the main WCSAxes class. So could you add these to a new ``astropy.visualization.wcsaxes.helpers`` submodule as functions and import them into the ``astropy.visualization.wcsaxes`` namespace?

astrofrog commented on astropy/visualization/wcsaxes/core.py:
```diff
@@ -307,25 +332,179 @@ def plot_coord(self, *args, **kwargs):
+
+        By default, this method will search for the ``BMAJ``, ``BMIN``, and ``BPA``
+        keywords in the FITS header to set the major and minor axes and the
+        position angle on the sky.
```
I don't think this should be mentioned here - rather, below in the ``header`` kwarg description, mention that if specified those keywords will be searched.

smaret replied:
Done in 3f903d5a1.

astrofrog commented on astropy/visualization/wcsaxes/core.py:
```diff
@@ -307,25 +332,179 @@ def plot_coord(self, *args, **kwargs):
+            Header containing the beam parameters.
+        major : float or :class:`~astropy.units.Quantity`, optional
+            Major axis of the beam in degrees or an angular quantity (overrides
+            ``BMAJ`` if present in header).
```
I think we should either allow header to be passed or major/minor/angle, not both.

smaret replied:
Done in f1e3d655c.

astrofrog commented on astropy/visualization/wcsaxes/core.py:
```diff
@@ -307,25 +332,179 @@ def plot_coord(self, *args, **kwargs):
+        **kwargs,
+    ):
+        """Add a scale bar
+
```
Mention that this only works for celestial WCS

astrofrog reviewed (changes requested):
Sorry for taking so long to get back to this - I've left some comments below to try and move forward - once these are addressed, I can help get the figure tests working properly (as we've recently refactored the infrastructure for that).

smaret commented:
Thanks for the comments @astrofrog. I'll look into this asap.

smaret commented:
@astrofrog I believe that I've addressed all your comments. I've also rebased  on `main` and modified the tests to use the new `@figure_test` decorator.  The tests seem to run fine, but `codecov`  seems to be confused with the image tests: it marks are "uncovered" lines of code that are indeed tested.

astrofrog commented:
@smaret - sorry to ask but can you rebase this on the latest ``main``, as ``CircleCI`` (where the figure tests run) was broken when you last pushed and should be fixed now?

smaret commented:
@astrofrog I've rebased on `main`, but the `CircleCI` test still fails...

smaret commented:
I finally managed to get the CircleCI test running after login into CircleCI from GitHub. The image tests are falling through, because the reference images are missing. The documentation says that [these images are created when the PR is merged,](https://docs.astropy.org/en/latest/development/testguide.html#generating-reference-images) so I guess I can ignore these failures.

smaret commented:
I've fixed the documentation build on Read the Doc with 83c1543. The reasons for this failure are still not completely clear to me though.

astrofrog commented:
Thanks for cleaning this up! I'll take a look at it shortly.

ayshih commented:
Re-iterating [my comment above](https://github.com/astropy/astropy/pull/12102#issuecomment-1056997224):

> It looks like these methods as currently written would be inaccurate when:
> 
> * The pixel scales at the reference pixel are different from the pixel scales within the image extent (e.g., when the reference pixel is well outside of the image extent and the projection is non-linear)
> * The pixel scales in the two directions are very different from each other (e.g., rectangular pixels)

That first bullet also applies to any situation where the projection is non-linear, as long as the reference pixel is not in the exact spot where the viewer is interested in (consider any full-sky projection).  I maintain that it'd be prudent to add corresponding notes to both of these functions' docstrings.

smaret commented:
> Re-iterating [my comment above](https://github.com/astropy/astropy/pull/12102#issuecomment-1056997224):

Thanks for the reminder. I've added a note in the docstrings with a1d98b0c.

ayshih commented:
> > Re-iterating [my comment above](https://github.com/astropy/astropy/pull/12102#issuecomment-1056997224):
> 
> Thanks for the reminder. I've added a note in the docstrings with [a1d98b0](https://github.com/astropy/astropy/commit/a1d98b0ce7401f4ef6aff819fa093c20d40a161c).

Thanks!

astrofrog reviewed (commented):
@smaret - thanks for the latest updates! Could you add a section in:

https://github.com/astropy/astropy/blob/main/docs/whatsnew/5.2.rst

to showcase the new functionality?

Also for the CircleCI failures, the figure hash files need to be updated, but I'm happy to take care of that and push to this PR once the What's new entry is there, and I'll then merge to get this into v5.2.

smaret commented:
Here is what the code snippet in the What's new file gives:

![showcase](https://user-images.githubusercontent.com/95471/196975065-41284495-6728-494b-bf6a-8cb5066698ee.png)

smaret commented:
Sorry for the multiple force-pushs, but I could not get the doctests to pass. It's working now.

astrofrog commented:
I've added the figure hashes for the tests and have defined ``__all__`` in helpers.py to avoid having to rename the imports. If CI passes, I'll approve and merge!

smaret commented:
pre-commit.ci autofix

smaret commented:
@astrofrog I think there is a problem with the [codecov check](https://github.com/astropy/astropy/pull/12102/checks?check_run_id=9027145326): it claims that the `helpers.py` file isn't covered by tests. Perhaps the codecov check on CI does not run figure tests?

astrofrog reviewed (approved):
Looks good to me! The coverage issue is because CircleCI doesn't currently upload coverage results, and the image tests only run if --remote-data=any is specified and there are no jobs that upload the coverage on GitHub Actions when --remote-data=any.

astrofrog commented:
@smaret - thank you for this contribution!
